### ISSUE TITLE
ArangoDB Nested/Join Query returns incorrect results

### ISSUE THREAD

**Issue by aveiros** [arangodb/arangodb#12247]
## My Environment

* __ArangoDB Version__:        3.5.3
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server / Active Failover
* __Deployment Strategy__:   docer
* __Configuration__:               n/a
* __Infrastructure__:               own
* __Operating System__:        MacOS
* __Total RAM in your machine__:        n/a
* __Disks in use__:        n/a
* __Used Package__:              Docker

## Component, Query & Data

__Affected feature__:
Arango SEARCH with Nested/Join Query

__Dataset__:
**docs** collection (type: document) with the following data
```
[
  {
    "_key": "a1",
    "_id": "docs/a1",
    "type": "a"
  },
  {
    "_key": "a2",
    "_id": "docs/a2",
    "type": "a"
  },
  {
    "_key": "a3",
    "_id": "docs/a3",
    "type": "a"
  },
  {
    "_key": "b1",
    "_id": "docs/b1",
    "type": "b"
  },
  {
    "_key": "b2",
    "_id": "docs/b2",
    "type": "b"
  }
]
```

**edges** collection (type: edge) with the following data
```
[
  {
    "_key": "l1",
    "_id": "edges/l1",
    "_from": "docs/a1",
    "_to": "docs/b1",
    "from": "a1",
    "to": "b1",
    "fromType": "a",
    "toType": "b"
  },
  {
    "_key": "l2",
    "_id": "edges/l2",
    "_from": "docs/b2",
    "_to": "docs/a2",
    "from": "b2",
    "to": "a2",
    "fromType": "b",
    "toType": "a"
  }
]
```

__Views__:
**docs_view** definition
```
{
  "links": {
    "docs": {
      "analyzers": [
        "identity"
      ],
      "fields": {},
      "includeAllFields": true,
      "storeValues": "id",
      "trackListPositions": false
    }
  }
}
```

**edges_view** definition
```
{
  "links": {
    "edges": {
      "analyzers": [
        "identity"
      ],
      "fields": {},
      "includeAllFields": true,
      "storeValues": "id",
      "trackListPositions": false
    }
  }
}
```

## Steps to reproduce 

1. Run query below
```
FOR aa IN (FOR a IN docs_view  SEARCH  a.type == "a" RETURN a._key)

    FOR ll IN (FOR l IN edges_view  SEARCH (l.fromType == "a" AND l.toType == "b") AND (l.from == aa OR l.to == aa) RETURN l.from == aa ? l.to : l.from)

        FOR bb IN (FOR b IN docs_view  SEARCH b.type == "b"  AND b._key == ll RETURN b._key)
    
RETURN DISTINCT aa
```

__Problem/Actual result__:
```
[
  "a1",
  "a2",
  "a3"
]
```

__Expected result__:
```
[
  "a1"
]
```

Running the following query, i get the correct results.
```
FOR aa IN [ "a1", "a2", "a3" ]

    FOR ll IN (FOR l IN edges_view  SEARCH (l.fromType == "a" AND l.toType == "b") AND (l.from == aa OR l.to == aa) RETURN l.from == aa ? l.to : l.from)

        FOR bb IN (FOR b IN docs_view  SEARCH b.type == "b"  AND b._key == ll RETURN b._key)
    
RETURN DISTINCT aa
```
```
[
  "a1"
]
```

Also running the following query with collections and FILTER, i get the correct results.
```
FOR aa IN (FOR a IN docs  FILTER  a.type == "a" RETURN a._key)

    FOR ll IN (FOR l IN edges  FILTER (l.fromType == "a" AND l.toType == "b") AND (l.from == aa OR l.to == aa) RETURN l.from == aa ? l.to : l.from)

        FOR bb IN (FOR b IN docs  FILTER b.type == "b"  AND b._key == ll RETURN b._key)
    
RETURN DISTINCT aa
```
```
[
  "a1"
]
```

Reproduced in 3.5.3, 3.6.4 and the nightly.

Let me know if you need more information. 😎

**Comment by aveiros** [comment]
Here's another example, where joins return incorrect results:

Dataset:
**docs** collection (type: document) with the following data

```
 [
   {
     "_key": "a1",
     "_id": "docs/a1",
     "type": "a"
   },
   {
     "_key": "a2",
     "_id": "docs/a2",
     "type": "a"
   },
   {
     "_key": "a3",
     "_id": "docs/a3",
     "type": "a"
   },
   {
     "_key": "b1",
     "_id": "docs/b1",
     "type": "b"
   },
   {
     "_key": "b2",
     "_id": "docs/b2",
     "type": "b"
   },
   {
     "_key": "b3",
     "_id": "docs/b3",
     "type": "b"
   },
   {
     "_key": "c1",
     "_id": "docs/c1",
     "type": "c"
   }
 ]
```

**edges** collection (type: edge) with the following data
```
[
   {
     "_key": "l1",
     "_id": "edges/l1",
     "_from": "docs/a1",
     "_to": "docs/b1",
     "subType": "xxx",
     "from": "a1",
     "to": "b1"
   },
   {
     "_key": "l2",
     "_id": "edges/l2",
     "_from": "docs/a1",
     "_to": "docs/b2",
     "subType": "xxx",
     "from": "a1",
     "to": "b2"
   },
   {
     "_key": "l3",
     "_id": "edges/l3",
     "_from": "docs/a2",
     "_to": "docs/c1",
     "subType": "yyy",
     "from": "a2",
     "to": "c1"
   },
   {
     "_key": "l4",
     "_id": "edges/l4",
     "_from": "docs/a3",
     "_to": "docs/b3",
     "subType": "xxx",
     "from": "a3",
     "to": "b3"
   }
 ]
```

__Query with bug__:
```
FOR b IN docs_view  SEARCH  b.type == "b"
 
    FOR eId IN ( FOR l IN edges_view  SEARCH l.subType == "xxx" AND  (l.from == b._key OR l.to == b._key)  RETURN l.from == b._key ? l.to : l.from )
    
    FOR a IN docs_view  SEARCH  a._key == eId  AND  a.type == "a"
    
     RETURN DISTINCT { a: a._key, b: b._key }
```

__Problem/Actual result__:
```
[
  {
    "a": "a1",
    "b": "b1"
  },
  {
    "a": "a1",
    "b": "b2"
  },
  {
    "a": "a1",
    "b": "b3"
  }
]
```

__Expected result__:
```
[
  {
    "a": "a1",
    "b": "b1"
  },
  {
    "a": "a1",
    "b": "b2"
  },
  {
    "a": "a3",
    "b": "b3"
  }
]
```

If we introduce the DISTINCT keyword on the query as follow, it hides the problem in a way... but sometimes we want duplicates.

__Query with DISTINCT__:
```
FOR b IN docs_view  SEARCH  b.type == "b"
 
    FOR eId IN ( FOR l IN edges_view  SEARCH l.subType == "xxx" AND  (l.from == b._key OR l.to == b._key)  RETURN DISTINCT l.from == b._key ? l.to : l.from )
    
    FOR a IN docs_view  SEARCH  a._key == eId  AND  a.type == "a"
    
     RETURN DISTINCT { a: a._key, b: b._key }
```

__Result__:
```
[
  {
    "a": "a1",
    "b": "b1"
  },
  {
    "a": "a3",
    "b": "b3"
  }
]
```

**Comment by aveiros** [comment]
The following is also occurring...

Without distinct
```
 FOR a IN docs_view  SEARCH  a.type == "a"
 
   FOR eId IN ( FOR l IN edges_view  SEARCH  l.subType == "xxx" AND  (l.from == a._key OR l.to == a._key)  RETURN l.from == a._key ? l.to : l.from ) 
 
     FOR b IN docs_view  SEARCH  b._key == eId  AND  b.type == "b"
 
 RETURN { a: a._key, b: b._key }
```

```
[
  {
    "a": "a1",
    "b": "b1"
  },
  {
    "a": "a1",
    "b": "b1"
  },
  {
    "a": "a3",
    "b": "b1"
  }
]
```

**Comment by Dronplane** [comment]
Hi @aveiros !
We are still investigating the issue. Meanwhile you could use the following workaround:
Please add `{ optimizer: { rules: [ "+inline-subqueries" ] } }`  to your query options. Disabling "inline-subqueries" rule  will fix query execution.

**Comment by joerg84** [comment]
> Hi @aveiros !
> We are still investigating the issue. Meanwhile you could use the following workaround:
> Please add `{ optimizer: { rules: [ "+inline-subqueries" ] } }` to your query options. Disabling "inline-subqueries" rule will fix query execution.

P.S. Note that this will have impact on the subquery performance, and as said we are looking into the proper fix.

**Comment by aveiros** [comment]
Thanks, i will give it a go this afternoon.

**Comment by Dronplane** [comment]
One note:
Not "+inline-subqueries" but of course "-inline-subqueries". Fixed in original comment also.

**Comment by aveiros** [comment]
Wrote a few tests with the rule `-inline-subqueries` against the examples provided... all testes passed.

**Comment by maxkernbach** [comment]
Closing as fixed with 3.5.6 and 3.7.2.

> * Fixed issue ES-664: the optimizer rule `inline-subqueries` must pull out a
>   subqueries that contains a COLLECT statement if the subquery is itself called
>   from within a loop. Otherwise the COLLECT will be applied to the values in the
>   outer FOR loop, which can produce a different result.
